packages/next/client/page-loader.js

```diff
@@ -70,14 +69,54 @@ export default class PageLoader {
         return
       }
 
-      // Load the script if not asked to load yet.
       if (!this.loadingRoutes[route]) {
-        this.loadScript(route)
+        // Make sure we don't load a dependency that's already loaded
+        const loadedModules = new Set(
+          Array.from(document.getElementsByTagName('script'), el => {
+            const results = /\/_next\/.*/g.exec(el.src)
+            return results ? results[0] : el.src
+          })
+        )
+        const deps = this.getDependencies(route)
+        const urlsToLoad = deps
+          .map(url => '/_next/' + url)
+          .filter(url => !loadedModules.has(url))
+
+        // Wait for all of the page's dependencies to load before loading
+        // the main page script
+        Promise.all(urlsToLoad.map(this.loadDependency)).then(() => {
+          this.loadScript(route)
+        })
         this.loadingRoutes[route] = true
       }
     })
   }
 
+  // Retrieve a list of dependencies for a given route from the build manifest
+  getDependencies (route) {
+    if (!window.__BUILD_MANIFEST) {
```

Comment:
I've gone back and changed the manifest naming algorithm to use hashes instead of concatenated filenames, which should cut down on the size of the manifest by a lot. A possible further optimization would be to expose the build manifest through an API, so the client could request the manifest for a particular page, rather than the whole thing. I'd like to get some performance numbers on this implementation before deciding whether to pursue that optimization.

Reply:
Agreed, I think just having a rough napkin calculation here would be more useful in the short term than actually exploring size mitigation strategies.

Reply:
Sure. For a rough calculation, an average module filepath with a hash filename looks like `"static/chunks/AAvCPKHKrG16L3HW4ENY8BZatII=.c727c892b8a902e00867.js",` and is approximately 70 characters long. That's virtually all that's in the manifest, so the base size should be something like `70 Bytes * number of pages * avg number of dependencies per page`.

Note that these files compress very well, since there's a lot of repeated content. For instance, on my small test app, the client manifest is 2.2kb on disk, and ~400 bytes gzipped. That ratio has actually gone down a bit since I switched over to the hashes--presumably because the filename-based bundle names had more repetition. I'm going to look into shortening the hashes.